### PR TITLE
feat(linter): improve no_redeclare rule implementation

### PR DESCRIPTION
**Pull request by Dunqing**

### CODE REVIEW

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #2084** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2084?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2084?utm_source=stack-comment).

File: crates/oxc_linter/src/rules/eslint/no_redeclare.rs [hunk 1]
```diff
@@ -58,7 +58,7 @@ declare_oxc_lint!(
     /// var a = 10;
     /// ```
     NoRedeclare,
-    nursery // There are false positives within TypeScript files (e.g. redeclare on interface)
+    pedantic
```
Comment (Dunqing) [inline crates/oxc_linter/src/rules/eslint/no_redeclare.rs:61]: We should check `FunctionScopedVariable` only

File: crates/oxc_linter/src/rules/eslint/no_redeclare.rs [hunk 2]
```diff
@@ -153,12 +157,10 @@ fn test() {
...
         ("export var a; var a;", None),
-        // `var` redeclaration in class static blocks. Redeclaration of functions is not allowed in class static blocks.
         ("class C { static { var a; var a; } }", None),
-        // Todo: Fix me
-        // ("class C { static { var a; { var a; } } }", None),
-        // ("class C { static { { var a; } var a; } }", None),
-        // ("class C { static { { var a; } { var a; } } }", None),
+        ("class C { static { var a; { var a; } } }", None),
+        ("class C { static { { var a; } var a; } }", None),
+        ("class C { static { { var a; } { var a; } } }", None),
```
Comment (Dunqing) [inline crates/oxc_linter/src/rules/eslint/no_redeclare.rs:163]: These tests already passed before this pr

File: crates/oxc_semantic/src/builder.rs [hunk 3]
```diff
@@ -253,7 +252,9 @@ impl<'a> SemanticBuilder<'a> {
     ) -> SymbolId {
         if let Some(symbol_id) = self.check_redeclaration(scope_id, span, name, excludes, true) {
             self.symbols.union_flag(symbol_id, includes);
-            self.add_redeclared_variables(VariableInfo { name: name.clone(), span, symbol_id });
+            if includes.is_function_scoped_declaration() {
```
Comment (Dunqing) [inline crates/oxc_semantic/src/builder.rs:255]: This rule focuses only on FunctionScopeVariable, e.g., a variable or function parameter defined via `var`.

**Comment by Dunqing** [comment]
image

We still get a 4% performance improvement in the pdf.js benchmark

**Comment by Boshen** [comment]
Can you test this in other codebases?

I assume you have a folder of codebases to test lint rules :-), if not:

https://github.com/oxc-project/oxc/blob/main/MAINTENANCE.md#release-linter

`oxlint -D no-redeclare`

**Comment by Boshen** [comment]
Feel free to merge once you have tested this PR in real codebases.

**Comment by Dunqing** [comment]
image

The https://github.com/oxc-project/oxc/pull/2010 caused a panic.

**Comment by Dunqing** [comment]
I have tested it, and the report message is expected. I think we can good to merge it.

**Pull request merged**